Implement XEP-0444: Message Reactions in Smack

ismael221 opened the pull request:
This commit adds support for XEP-0444 (Message Reactions) in Smack. Key changes include:

- Added ReactionsManager to handle reactions, including adding, removing, and listening for reactions on messages.
- Introduced ReactionsElement and Reaction classes to represent the <reactions> element and individual emoji reactions.
- Added ReactionsFilter to detect messages containing reactions.
- Implemented ReactionRestrictions to manage restrictions like max reactions per user and allowed emojis.
- Integrated reaction restrictions with XMPP service discovery.
- Added ReactionsListener for applications to handle incoming reactions.
- Included unit tests to verify functionality.

This enables emoji reactions in XMPP messages, with support for restrictions and service discovery.

Related: XEP-0444 (https://xmpp.org/extensions/xep-0444.html)

Thank you for your contribution!

Before creating a Pull Request, please make sure to
* read https://github.com/igniterealtime/Smack/wiki/Guidelines-for-Smack-Developers-and-Contributors
* run `gradle check` successfully in order to make sure that your code does not break any JUnit tests and is conform to the projects code style.
* (if applicable) mention any Jira issue codes (eg. `SMACK-XXX`) in the *body* of your commit message (not the header), so that Jira automatically links the PR to the issue.
* squash your commits if possible/sensible.

Flowdalic commented on smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/ReactionsManager.java:
```diff
@@ -0,0 +1,357 @@
+    public static DataForm createReactionRestrictionsForm(int maxReactionsPerUser, List<String> allowedEmojis) {
+
+        DataForm.Builder builder = DataForm.builder();
+        builder.setFormType(String.valueOf(DataForm.Type.result));
```
That `String.valueOf` can probably be dropped.

ismael221 replied:
I couldn't drop it, as it requires a String

`   public Builder setFormType(String formType) {
            FormField formField = FormField.buildHiddenFormType(formType);
            return addField(formField);
        }`

Flowdalic replied:
That's because `DataForm.Type.result` is the wrong value. Look at XEP-0440 again what the right form type is.

Flowdalic commented on smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/ReactionsManager.java:
```diff
@@ -0,0 +1,357 @@
+     * @param listener The reactions listener to be added.
+     */
+    public synchronized void addReactionsListener(ReactionsListener listener){
+        listeners.add(listener);
```
We usually expose the return value of `add()`

Flowdalic commented on smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/ReactionsManager.java:
```diff
@@ -0,0 +1,357 @@
+     * @param listener The reactions listener to be removed.
+     */
+    public synchronized void removeReactionsListener(ReactionsListener listener){
+        listeners.remove(listener);
```
We usually expose the return value of `remove()`

Flowdalic commented on smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/ReactionsManager.java:
```diff
@@ -0,0 +1,357 @@
+        FormField.Builder<TextSingleFormField, TextSingleFormField.Builder> allowlistFieldBuilder = FormField.builder("allowlist");
+        for (String emoji : allowedEmojis) {
+            Reaction reaction = new Reaction(emoji);
+            FormField.builder("value").setValue((CharSequence) reaction);
```
This cast to `CharSequence` probably does not what you want.

Flowdalic commented on smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/provider/ReactionsElementProvider.java:
```diff
@@ -0,0 +1,52 @@
+        while (true) {
+            XmlPullParser.Event tag = parser.next();
+
+            if (tag == XmlPullParser.Event.END_ELEMENT && parser.getName().equals(ReactionsElement.ELEMENT)) {
```
Please follow the idiom in the other Providers where we break out of the lop if `END_ELEMENT` and depth are correct.

Flowdalic commented on smack-experimental/src/test/java/org/jivesoftware/smackx/reactions/ReactionTest.java:
```diff
@@ -0,0 +1,164 @@
+
+        ReactionsManager.addReactionsToMessage(message, emojis, messageId, null);
+
+        ReactionsElement reactionsElement = (ReactionsElement) message.getExtensionElement(ReactionsElement.ELEMENT, ReactionsElement.NAMESPACE);
```
Casts are bad, better use the type safe approach (may need adjustments)
```suggestion
        ReactionsElement reactionsElement = message.getExtensionElement(ReactionsElement.class);
```

Flowdalic commented on smack-experimental/src/test/java/org/jivesoftware/smackx/reactions/ReactionTest.java:
```diff
@@ -0,0 +1,164 @@
+        reactionsManager.reactionsElementListener(message);
+
+        // Assertions: Ensure that the message contains the reactions element
+        assertNotNull(message.getExtensionElement(ReactionsElement.ELEMENT, ReactionsElement.NAMESPACE));
```
```suggestion
        assertNotNull(message.getExtensionElement(ReactionsElement.class));
```

Flowdalic reviewed (changes requested):
Thanks for your contribution. Gave your PR a quick review from the top of my head. There is probably more.

ismael221 commented:
Hi Flowdalic! I've made all the requested changes from the previous review and addressed the feedback provided. Additionally, the code passed the gradlew check after the updates. Could you please review the PR again? Thank you for your attention!"

Flowdalic commented on smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/element/Reaction.java:
```diff
@@ -0,0 +1,106 @@
+public class Reaction implements ExtensionElement {
+
+    public static final String ELEMENT = "reaction";
+    public static final String NAMESPACE = StreamOpen.CLIENT_NAMESPACE;
```
That's not correct, the namespace of `<reaction/>` is not `StreamOpen.CLIENT_NAMESPACE`.

ismael221 replied:
I used this because the XEP specification doesn't have a namespace, but if I don't add a namespace, the tests fail and I can't parse it. In this case, should I correct the implementation to not use ReactionElement or is there another namespace to put in cases that are empty? because even an empty string is not accepted

Flowdalic commented on smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/element/Reaction.java:
```diff
@@ -0,0 +1,106 @@
+     * @param message The XMPP message from which to extract the reaction.
+     * @return The Reaction extension from the message, or {@code null} if not present.
+     */
+    public static Reaction fromMessage(Message message) {
```
I don't think we need this method.

Flowdalic commented on smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/element/ReactionsElement.java:
```diff
@@ -0,0 +1,123 @@
+
+        xml.rightAngleBracket();
+        for (Reaction reaction : reactions) {
+            if (reaction != null) {
```
Shouldn't be possible that `reactions` contains `null` values and this should be an invariant.

Flowdalic commented on smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/filter/ReactionsFilter.java:
```diff
@@ -0,0 +1,39 @@
+ * @author Ismael Nunes Campos
+ *
+ */
+public final class ReactionsFilter extends StanzaExtensionFilter {
```
I don't see the value of this class if the same could be achieved by `ExtensionElementFilter(ReactionsElement.class)`.

Flowdalic commented on smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/provider/ReactionsElementProvider.java:
```diff
@@ -0,0 +1,69 @@
+        outerloop: while (true) {
+            XmlPullParser.Event tag = parser.next();
+
+            if (tag == XmlPullParser.Event.START_ELEMENT) {
```
switch/case here please

Flowdalic commented on smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/provider/ReactionsElementProvider.java:
```diff
@@ -0,0 +1,69 @@
+                    reactions.add(new Reaction(emoji));
+                }
+            } else if (tag == XmlPullParser.Event.END_ELEMENT) {
+                if (parser.getName().equals(ReactionsElement.ELEMENT) && parser.getDepth() == initialDepth) {
```
The `parser.getName().equals(ReactionsElement.ELEMENT)` is unnecessary

Flowdalic commented on smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/ReactionsManager.java:
```diff
@@ -0,0 +1,364 @@
+        DiscoverInfo discoverInfo = sdm.discoverInfo(jid);
+
+
+        for (XmlElement extension : discoverInfo.getExtensions()) {
```
Use `getExtensions(DataForm.class)`, this also avoids the instanceof check in the line below

Flowdalic commented on smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/ReactionsManager.java:
```diff
@@ -0,0 +1,364 @@
+     * @throws InterruptedException If the operation is interrupted.
+     * @throws SmackException.NoResponseException If no response is received from the server.
+     */
+    public ReactionRestrictions getReactionRestrictions(EntityBareJid jid) throws XMPPException.XMPPErrorException, SmackException.NotConnectedException,
```
`EntityBareJId` here is wrong, you want to be able to query all XMPP addresses here (at least, that is my interpretation of the XEP).

Flowdalic commented on smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/ReactionsManager.java:
```diff
@@ -0,0 +1,364 @@
+        DataForm restrictionsForm = createReactionRestrictionsForm(maxReactionsPerUser, allowedEmojis);
+        sdm.addExtendedInfo(restrictionsForm);
+
+        sdm.addFeature(ReactionsElement.NAMESPACE);
```
This is already done in the class' constructor, why do it again here?

Flowdalic commented on smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/ReactionsManager.java:
```diff
@@ -0,0 +1,364 @@
+        builder.setFormType(String.valueOf(DataForm.Type.result));
+
+        builder.addField(
+                        FormField.builder("max_reactions_per_user").setValue(String.valueOf(maxReactionsPerUser))
```
I see `FormField.Builder.setValue(int)` in Smack, so the `String.valueOf` appears unnecessary

Flowdalic commented on smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/ReactionsManager.java:
```diff
@@ -0,0 +1,364 @@
+
+        FormField.Builder<TextSingleFormField, TextSingleFormField.Builder> allowlistFieldBuilder = FormField.builder("allowlist");
+        for (String emoji : allowedEmojis) {
+            Reaction reaction = new Reaction(emoji);
```
Why do you wrap the emoji in a `Reaction` just to extract the emoji again one line later?

Flowdalic commented on smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/ReactionsManager.java:
```diff
@@ -0,0 +1,364 @@
+     * @return                   A message builder containing the reactions to the original message.
+     * @throws IllegalArgumentException If the number of reactions exceeds the allowed limit or if any emoji is not allowed.
+     */
+    public MessageBuilder createMessageWithReactions(List<String> emojis, String originalMessageId, ReactionRestrictions restrictions) {
```
Any reason why this method is not `static`?

Flowdalic commented on smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/ReactionsManager.java:
```diff
@@ -0,0 +1,364 @@
+         *
+         * @return The list of allowed emojis.
+         */
+        public List<String> getAllowedEmojis() {
```
This should return a `Set` for fast lookup

Flowdalic commented on smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/ReactionsManager.java:
```diff
@@ -0,0 +1,364 @@
+     */
+    public static class ReactionRestrictions {
+        private final int maxReactionsPerUser;
+        private final List<String> allowedEmojis;
```
This should be a `Set`

Flowdalic commented on smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/ReactionsManager.java:
```diff
@@ -0,0 +1,364 @@
+        for (XmlElement extension : discoverInfo.getExtensions()) {
+            if (extension instanceof DataForm) {
+                DataForm dataForm = (DataForm) extension;
+                FormField formTypeField = dataForm.getField("FORM_TYPE");
```
Use `DataForm.getFormType()` to select the right form

Flowdalic commented on smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/element/Reaction.java:
```diff
@@ -0,0 +1,106 @@
+ * @see ExtensionElement
+ * @see Message
+ */
+public class Reaction implements ExtensionElement {
```
This should probably extend `NamedElement`

Flowdalic commented on smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/ReactionsManager.java:
```diff
@@ -0,0 +1,364 @@
+     *
+     * @param packet The received XMPP stanza (message).
+     */
+    public void reactionsElementListener(Stanza packet) {
```
Why is this method `public`?

Flowdalic replied:
Why did you mark things as resolved if they are not?

Flowdalic commented on smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/ReactionsManager.java:
```diff
@@ -0,0 +1,364 @@
+    public void reactionsElementListener(Stanza packet) {
+        Message message = (Message) packet;
+        ReactionsElement reactionsElement = ReactionsElement.fromMessage(message);
+        if (reactionsElement != null) {
```
That should never happen, otherwise, the filter wouldn't work properly.

Flowdalic commented on smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/ReactionsManager.java:
```diff
@@ -0,0 +1,364 @@
+     */
+    public ReactionsManager(XMPPConnection connection) {
+        super(connection);
+        connection.addAsyncStanzaListener(this::reactionsElementListener, reactionsElementFilter);
```
This stanza listener shouldn't only be setup if there are actual reaction listeners.

ismael221 replied:
I don't think that I fully understand what you mean

ismael221 replied:
Would this be a good approach?


`  public ReactionsManager(XMPPConnection connection) {
        super(connection);

        connection.addAsyncStanzaListener(new StanzaListener() {
            @Override public void processStanza(Stanza packet)
                            throws SmackException.NotConnectedException, InterruptedException,
                            SmackException.NotLoggedInException {
                final Message message = (Message) packet;
                final ReactionsElement reactionsElement = ReactionsElement.fromMessage(message);
                final Set<Reaction> reactionsSet = reactionsElement.getReactions();
                final Set<String> reactionsStrings = new HashSet<>();
                
                for (Reaction reaction : reactionsSet) {
                    reactionsStrings.add(reaction.getEmoji());
                }

                asyncButOrdered.performAsyncButOrdered(message.getFrom().asBareJid(), () -> {
                    for (ReactionsListener l: listeners){
                        l.onReactionReceived(message.getFrom().asBareJid().toString(),reactionsStrings,reactionsElement,message);
                    }
                });
            }
        }, REACTIONS_FILTER);

        serviceDiscoveryManager = ServiceDiscoveryManager.getInstanceFor(connection);
    }`

Flowdalic replied:
> Would this be a good approach?

No, only add the stanza listener once the first reaction listener is added, and remove the stanza listener once the last reaction listener got removed.

Flowdalic commented on smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/ReactionsManager.java:
```diff
@@ -0,0 +1,364 @@
+     * @param message The XMPP message that contains the reactions.
+     * @param reactionsElement The {@link ReactionsElement} containing the reactions.
+     */
+    public void notifyReactionListeners(Message message, ReactionsElement reactionsElement) {
```
Why is this method `public`? Why does it exist at all if there is only exactly one call site?

Flowdalic commented on smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/ReactionsManager.java:
```diff
@@ -0,0 +1,364 @@
+     */
+    public void notifyReactionListeners(Message message, ReactionsElement reactionsElement) {
+        for (ReactionsListener listener : listeners) {
+            asyncButOrdered.performAsyncButOrdered(message.getFrom().asBareJid(), () -> {
```
Using `asyncButOrdered` here is too late, since the initial listener was already async the messages for a particular jid could be already out of order.

Flowdalic commented on smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/ReactionsManager.java:
```diff
@@ -0,0 +1,364 @@
+        super(connection);
+        connection.addAsyncStanzaListener(this::reactionsElementListener, reactionsElementFilter);
+        ServiceDiscoveryManager sdm = ServiceDiscoveryManager.getInstanceFor(connection);
+        sdm.addFeature(ReactionsElement.NAMESPACE);
```
I don't think it is sensible to announce the feature right away. What if the Smack user did not setup a listener or isn't even aware of reactions? Then Smack would still announce the feature signalling to others that I can understand reactions.
A typical Smack idiom would be to only start announcing the feature once a reaction listener has been setup.

Flowdalic commented on smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/ReactionsListener.java:
```diff
@@ -0,0 +1,32 @@
+     * @param reactionsElement reactionsElement
+     * @param message message
+     */
+    void onReactionReceived(Message message, ReactionsElement reactionsElement);
```
Let's make the API of `ReactionsListener` more user friendly. The user is primary interested in two aspects: the ID of the message that was reacted upon, and the reactions itself. Everything else is secondary.
```suggestion
    void onReactionReceived(String id, List<String> reactions, ReactionsElement reactionsElement, Message message);
```

Also make sure to document what `message` is. Is it the Message the reaction arrived with, or is the the message on which was reacted on?

Flowdalic replied:
Why was this marked as resolved if it wasn't resolved?

Flowdalic commented on smack-experimental/src/test/java/org/jivesoftware/smackx/reactions/ReactionTest.java:
```diff
@@ -0,0 +1,169 @@
+     * @throws SmackParsingException If a failure occurs while parsing the XML.
+     */
+    @Test
+    void testReaction() throws XmlPullParserException, IOException, SmackParsingException {
```
All parsing related tests should be parameterized:
```
    @ParameterizedTest
    @EnumSource(SmackTestUtil.XmlPullParserKind.class)
```

Flowdalic commented on smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/provider/ReactionsElementProvider.java:
```diff
@@ -0,0 +1,69 @@
+
+    @Override
+    public ReactionsElement parse(XmlPullParser parser, int initialDepth, XmlEnvironment xmlEnvironment) throws XmlPullParserException, IOException, SmackParsingException, ParseException {
+        String id = parser.getAttributeValue(null, "id");
```
```suggestion
        String id = parser.getAttributeValue("id");
```

Flowdalic commented on smack-experimental/src/test/java/org/jivesoftware/smackx/reactions/ReactionTest.java:
```diff
@@ -0,0 +1,169 @@
+        // Act: Call the listener
+        XMPPConnection connection = mock(XMPPConnection.class); // Mock of XMPP connection
+        ReactionsManager reactionsManager = new ReactionsManager(connection);
+        reactionsManager.reactionsElementListener(message);
```
How does any of this test the reactions element listener?

Flowdalic commented on smack-experimental/src/test/java/org/jivesoftware/smackx/reactions/ReactionTest.java:
```diff
@@ -0,0 +1,169 @@
+    }
+
+    @Test
+    void testXMLParsing() {
```
This does not test parsing, this tests XML serialization.

ismael221 commented:
Ready for review
